### PR TITLE
E2E Test Pipeline Refactoring

### PR DESCRIPTION
**Pull request by AhmedBM**
## Description

* Refactored stages into templates:
  * Allows to easily scale targets be reusing templates (See stages folder below).
  * Allows retry failed targets, in case of timeouts or any other random failures without needing a new run.
  * Base pipelines much smaller and easier to understand (400 lines vs 120)
* Ability to target specific tests (Test Expression)
* Added ability to throw dynamic targets at the pipelines, by default contains:

```
    - distroName: ubuntu-18.04
      image_publisher: Canonical
      image_offer: UbuntuServer
      image_sku: 18.04-LTS
      image_version: latest
      packagePattern: '**/*bionic_x86_64.deb'
      device_id: ubuntu1804
      pre_script: >-
        export DEBIAN_FRONTEND=noninteractive &&
        curl https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > microsoft.gpg &&
        sudo cp ./microsoft.gpg /etc/apt/trusted.gpg.d/ &&
        curl https://packages.microsoft.com/config/ubuntu/18.04/multiarch/prod.list > ./microsoft-prod.list &&
        sudo cp ./microsoft-prod.list /etc/apt/sources.list.d/ &&
        sudo apt update &&
        sudo apt install aziot-identity-service
    - distroName: ubuntu-20.04
      image_publisher: Canonical
      image_offer: 0001-com-ubuntu-server-focal
      image_sku: 20_04-lts-gen2
      image_version: latest
      packagePattern: '**/*focal_x86_64.deb'
      device_id: ubuntu2004
      pre_script: >-
        export DEBIAN_FRONTEND=noninteractive &&
        curl https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > microsoft.gpg &&
        sudo cp ./microsoft.gpg /etc/apt/trusted.gpg.d/ &&
        curl https://packages.microsoft.com/config/ubuntu/20.04/prod.list > ./microsoft-prod.list &&
        sudo cp ./microsoft-prod.list /etc/apt/sources.list.d/ &&
        sudo apt update &&
        sudo apt install aziot-identity-service
```
## Checklist

- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [X] All unit tests are passing.
- [X] I have merged the latest `main` branch prior to this PR submission.
- [X] I submitted this PR against the `main` branch.

### CODE REVIEW

File: devops/pipeline/stages/e2etest_cloud.yaml [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  type: object
+- name: skipTeardown
+  displayName: Leave test infrastructure in-place
+  type: boolean
+  default: false
+- name: twinTimeout
+  displayName: Timeout for twin updates
+  type: number
+  default: 90
+# - name: PRE_SCRIPT
```
Comment (robertschaedler3) [inline devops/pipeline/stages/e2etest_cloud.yaml:33]: Can we remove this pre-script if it is no longer used?

File: devops/pipeline/stages/e2etest_cloud.yaml [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+#   type: string
+#   default: ' '
+- name: postScript
+  displayName: 'Shell script to execute after osconfig install'
+  type: string
+  default: ' '
+- name: testExpression
+  type: string
+  default: ' '
+- name: sourceBranch
```
Comment (robertschaedler3) [inline devops/pipeline/stages/e2etest_cloud.yaml:44]: Is this different than the branch that the E2E Pipeline run is triggered on? Shouldn't the sources always be taken from the package build on the same branch?

File: devops/scripts/install_deps_ubuntu2004.sh [hunk 3]
```diff
@@ -4,4 +4,7 @@ apt -y update
 apt-get -y install software-properties-common
 apt-add-repository ppa:lttng/stable-2.10 -y
 apt -y update
-apt-get -y install git cmake build-essential curl libcurl4-openssl-dev libssl-dev uuid-dev libgtest-dev libgmock-dev liblttng-ust-dev rapidjson-dev ninja-build wget
\ No newline at end of file
+apt-get -y install git cmake build-essential curl libcurl4-openssl-dev libssl-dev uuid-dev libgtest-dev libgmock-dev liblttng-ust-dev rapidjson-dev ninja-build wget
+# Install Dotnet - Needed for e2e tests
+# See https://docs.microsoft.com/en-us/dotnet/core/install/linux-ubuntu#1804-
+curl -sSL https://dot.net/v1/dotnet-install.sh | bash /dev/stdin
```
Comment (MariusNi) [inline devops/scripts/install_deps_ubuntu2004.sh:10]: Instructions for manually building OSConfig on Ubuntu 18.04 and 20.04 are identical. Why are these scripts here different? Can we have a single script for all Ubuntu targets?

File: devops/pipeline/stages/e2etest_cloud.yaml (hunk 2 above)
Comment (AhmedBM) [reply devops/pipeline/stages/e2etest_cloud.yaml:44]: yes, and thats why that is the default behavior. If you are testing new scenarios on existing binaries however (whats is already on main), this is where it comes in useful. In order to trigger the tests we require the topic branch to have packages available.

File: devops/scripts/install_deps_ubuntu2004.sh (hunk 3 above)
Comment (AhmedBM) [reply devops/scripts/install_deps_ubuntu2004.sh:10]: I can probably have one script that targets both distros, i mainly wanted to have the dotnet install included in the scripts, these scripts have not changed in awhile and are based on what is in the Dockerfile for those environments so i rather not touch it right now.

File: devops/pipeline/stages/e2etest_cloud.yaml (hunk 2 above)
Comment (robertschaedler3) [reply devops/pipeline/stages/e2etest_cloud.yaml:40]: Ahh I see, thanks for the explanation 👍🏻

**Pull request merged**